### PR TITLE
Reversegeocode

### PR DESCRIPTION
**Pull request by marhab21**
I added a Reverse Geocoder to the Custom Geocoder.html Sandcastle demo, and I also added a new demo, which is another Reverse Geocoder, using BingMaps, and giving the address every time the user clicks on a location on the map.
This will demonstrate various features of Cesium, including comparing the information retrieved from OpenStreetMap to the one retrieved from Bing Maps.
Let me know what documentation or what else I need to add, or what I might be doing wrong : )
I had to edit .gitignore in order to add a gallery item.

### CODE REVIEW

**Comment by cesium-concierge** [comment]
[Please sign the CLA before we review this PR.](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#contributor-license-agreement-cla)

Welcome to the Cesium community @marhab21!

Can you please send in a [Contributor License Agreement](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#contributor-license-agreement-cla) (CLA) so that we can review and merge this pull request?

:warning: I noticed that [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CHANGES.md) has not been updated. If this change updates the public API in any way, fixes a bug, or makes any non-trivial update, please add a bullet point to `CHANGES.md` and comment on this pull request so we know it was updated. For more info, see the [Pull Request Guidelines]( https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#pull-request-guidelines).

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

**Comment by marhab21** [comment]
I have already sent a CLA. I failed eslint, but corrected the errors.

**Comment by marhab21** [comment]
What do I do to get the code reviewed?

**Comment by marhab21** [comment]
Still says CLA Required...hmmm...I sent it to cla@agi.com...should I do something else?

**Comment by hpinkos** [comment]
Hi @marhab21, we're a bit busy this week, but someone should have time to review your pull request soon.  Thanks!

**Comment by marhab21** [comment]
Hi Hannah!
Thank you, no problem, I was just wondering if there was anything else I
needed to do to get a review...
No hurry :)

Thanks again,

Martine

**Comment by pjcozzi** [comment]
We now have a CLA from @marhab21, thanks again!

File: .gitignore [hunk 1]
```diff
@@ -10,7 +10,7 @@ Thumbs.db
 /Apps/CesiumViewer/Gallery/gallery-index.js
 
 /Apps/Sandcastle/jsHintOptions.js
-/Apps/Sandcastle/gallery/gallery-index.js
+#/Apps/Sandcastle/gallery/gallery-index.js
```
Comment (hpinkos) [inline .gitignore]: Can you revert this change?

Comment (hpinkos) [reply .gitignore]: We don't want to commit `gallery-index.js`.  It is generated automatically when you run `npm run build`

File: Apps/Sandcastle/gallery/Reverse Geocoder.html [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+ viewer.geocoder.viewModel.complete.addEventListener(function () {
+     if (!viewer.geocoder.viewModel.searchText.empty){
+         setPointForSearchLocation();
+     }
+ });
+
+ /**
+ * Find the camera position after search, and put a point there.
+ */
+  function setPointForSearchLocation() {
```
Comment (hpinkos) [inline Apps/Sandcastle/gallery/Reverse Geocoder.html]: Something in this function is causing an error to be thrown when you use the geocoder search

**Comment by hpinkos** [comment]
Thanks for the pull request @marhab21!  I think these demos will be helpful for people.
Those are all of the comments I have for now.  Be sure to add a comment to `CHANGES.md` about your change

**Comment by marhab21** [comment]
Yes, I will...sorry...thanks Hannah!

Martine

File: Apps/Sandcastle/gallery/Custom Geocoder.html [hunk 3]
```diff
@@ -32,28 +32,48 @@
...
 <script id="cesium_sandcastle_script">
 function startup(Cesium) {
     'use strict';
 //Sandcastle_Begin
 /**
- * This class is an example of a custom geocoder. It provides geocoding through the OpenStreetMap Nominatim service.
- * @alias OpenStreetMapNominatimGeocoder
- * @constructor
- */
+* This class is an example of a custom geocoder, and reverse geocoder. It provides both * * through the OpenStreetMap Nominatim service. Press the "Reverse Geocoder" button to go * * to OpenStreet Map headquarters. OpenStreetMap has more information in the UK than in the * US.
```
[the remark this answers is not in the document]
Comment (marhab21) [reply Apps/Sandcastle/gallery/Custom Geocoder.html]: No, I think it is a result of formatting when I got rid of the tabs...will fix : )

**Comment by marhab21** [comment]
OK, I hope I have fixed everything. Do I need to close the pull request, or do I wait 'til it's merged?
Thank you so much Hannah for the review...

**Comment by hpinkos** [comment]
You can just leave this PR open.  I'll take another look at it tomorrow.

Does anyone else want to look at these sandcastle examples? @pjcozzi?

**Comment by pjcozzi** [comment]
Sounds like a great example, but I have to pass due to time constraints.

**Comment by mramato** [comment]
While I'm fine with having a reverse geocoding example (though I haven't run it yet), I'm not sure updating the custom geocoder example to add reverse geocoding is a good idea, I think it confuses what is otherwise a solid simple example.  Since this is all just example code (no Cesium to changes at all) I think just adding the new example would be preferred.

**Comment by marhab21** [comment]
Hi Matthew!
Thank you...
Should I revert the change on the reverse part of the Custom Geocoder?
Let me know if there is anything I need to do!
Thanks again,

Martine

**Comment by ggetz** [comment]
@marhab21 Revert the changes in the Custom Geocoder example, and add the code to a new Sandcastle example. You should make a new file like `Custom Geocoder.html` and put the code there. Also create a thumbnail image like the other examples.

**Comment by mramato** [comment]
@ggetz there is already a new Sandcastle example, we just need to revert the custom geocoder example and that's it.  There's no reason to have two reverse geocoding examples.

If you could reset this branch to clean up the history, that would be great too; but if you're not sure how to do that, don't worry about it.

**Comment by marhab21** [comment]
Thank you! I will revert the Custom Geocoder...I am not sure about how to reset the branch. I will look into it, but might leave it alone...I will confirm one way or the other.

**Comment by marhab21** [comment]
Well, I have tried everything, but my builds keep failing, and I can't figure out why. I did catch 1 lint error, which I fixed. But the build keeps failing mainly in KML files, and things that don't seem to have anything to do with my changes. What am I missing? Here is the raw log...
[Errors.txt](https://github.com/AnalyticalGraphicsInc/cesium/files/1515703/Errors.txt)

**Comment by hpinkos** [comment]
No worries @marhab21.  We've had some problems with CI randomly failing lately.  I'll take a closer look when I have a minute to do another review here, but It doesn't seem related to your changes.

**Comment by marhab21** [comment]
Thank you Hannah!! : )

**Comment by hpinkos** [comment]
This is a super cool demo, thanks for the contribution @marhab21!  I just pushed a commit with some code cleanup so your example follows the same [coding guidelines](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/CodingGuide) as our other demos.  I also removed the code to put a point at the location when you enter an address into the geocoder search bar because it wasn't as relevant to this example.  

Congratulations on your first contribution!

**Comment by hpinkos** [comment]
Just noticed that the formatting for `CHANGES.md` got a little messed up.  I fixed it in master: 2a21744

**Comment by marhab21** [comment]
Thank you everyone! I am so happy to have a contribution in Cesium!!
: )

**Comment by marhab21** [comment]
I found it in "Showcases", but it doesn't seem to load, and redirects to HelloWorld :(

**Comment by mramato** [comment]
> When will the demo be up in Official Sandcastle?

It will go out on Friday with the next Cesium release (Cesium releases on the first business day of every month).

**Comment by marhab21** [comment]
Awesome! Thank you so much!!

Martine
